File: pipeline/postprocess.py
"""Post-processing for the bump pipeline:

1. Pre-compensation (unsharp mask) to counteract the Eufy E1 printer's slicing
   smoothing.  Y = X + α (X − Gauss_σ(X)).
2. Quantization to 8-bit or 16-bit PNG bytes for the printer's slicer.

Input convention: bump map is float, centered around 0, roughly in [-0.5, 0.5].
Output convention: PNG bytes (uint8 mode 'L' or uint16 mode 'I;16').
"""

from __future__ import annotations

import io

import numpy as np
import cv2
from PIL import Image


def precompensate_for_printer(
    bump: np.ndarray,
    sigma: float = 1.0,
    alpha: float = 0.5,
) -> np.ndarray:
    """Apply unsharp masking sized to the printer's expected smoothing kernel.

    The Eufy Make E1 (with smoothing="1") slightly low-passes the depth map
    during slicing. By pre-amplifying the high-frequency component, we cancel
    a portion of that loss so the printed surface still carries fine detail.

    Parameters
    ----------
    bump : np.ndarray
        2D float bump map (any range; recommended [-0.5, 0.5]).
    sigma : float
        Gaussian sigma in pixels matching the printer's PSF.  σ→0 disables.
    alpha : float
        Pre-emphasis gain. 0 = no compensation; typical 0.3-0.7.

    Returns
    -------
    np.ndarray of float32
        Sharpened bump in approximately the same range as the input.
    """
    if bump.ndim != 2:
        raise ValueError(f"bump must be 2D, got shape {bump.shape}")
    bump = bump.astype(np.float32)

    if alpha == 0.0 or sigma <= 0.0:
        return bump.copy()

    # cv2 picks a kernel size from sigma when ksize=(0,0).
    blurred = cv2.GaussianBlur(bump, (0, 0), float(sigma))
    out = bump + float(alpha) * (bump - blurred)
    return out.astype(np.float32)


def quantize_to_png_bytes(bump: np.ndarray, bit_depth: int = 16) -> bytes:
    """Convert a centered float bump map to PNG bytes (8-bit 'L' or 16-bit 'I;16').

    The float input is shifted from [-0.5, 0.5] to [0, 1] (clamped first to
    [-0.5, 0.5]) and then scaled to the integer range.

    Parameters
    ----------
    bump : np.ndarray
        2D float array, expected centered in [-0.5, 0.5].
    bit_depth : int
        Either 8 or 16.

    Returns
    -------
    bytes
        PNG-encoded image bytes.
    """
    if bump.ndim != 2:
        raise ValueError(f"bump must be 2D, got shape {bump.shape}")
    if bit_depth not in (8, 16):
        raise ValueError(f"bit_depth must be 8 or 16, got {bit_depth}")

    clipped = np.clip(bump.astype(np.float32), -0.5, 0.5)
    shifted = clipped + 0.5  # -> [0, 1]
    if bit_depth == 16:
        arr = np.clip(shifted * 65535.0 + 0.5, 0, 65535).astype(np.uint16)
        # Use frombytes for explicit I;16 construction (avoids the deprecated
        # `mode=` parameter on Image.fromarray for type conversion).
        img = Image.new("I;16", (arr.shape[1], arr.shape[0]))
        img.frombytes(arr.tobytes())
    else:
        arr = np.clip(shifted * 255.0 + 0.5, 0, 255).astype(np.uint8)
        img = Image.fromarray(arr, mode="L")

    buf = io.BytesIO()
    img.save(buf, format="PNG")
    return buf.getvalue()


def quantize_to_array(bump: np.ndarray, bit_depth: int = 16) -> np.ndarray:
    """Same scaling logic as :func:`quantize_to_png_bytes` but returns the
    raw integer ndarray (uint8 or uint16). Useful for tests and the simulated
    print preview endpoint.
    """
    if bit_depth not in (8, 16):
        raise ValueError(f"bit_depth must be 8 or 16, got {bit_depth}")
    clipped = np.clip(bump.astype(np.float32), -0.5, 0.5)
    shifted = clipped + 0.5
    if bit_depth == 16:
        return np.clip(shifted * 65535.0 + 0.5, 0, 65535).astype(np.uint16)
    return np.clip(shifted * 255.0 + 0.5, 0, 255).astype(np.uint8)

File: web/src/postprocess.js
// Unsharp 预补偿 + 8/16-bit 量化（PNG 数据生产侧）。
//
// 移植自 pipeline/postprocess.py。
// 输入约定：bump 是 float，居中在 0，大致在 [-0.5, 0.5]。

import { gaussianBlur } from './filters.js';

/**
 * Y = X + α (X − Gauss_σ(X))
 *
 * @param {Float32Array} bump
 * @param {number} W
 * @param {number} H
 * @param {object} [opts]
 * @param {number} [opts.sigma=1.0]
 * @param {number} [opts.alpha=0.5]
 * @returns {Float32Array}
 */
export function precompensateForPrinter(bump, W, H, opts = {}) {
    const sigma = opts.sigma ?? 1.0;
    const alpha = opts.alpha ?? 0.5;
    if (alpha === 0.0 || sigma <= 0.0) return bump.slice();

    const blurred = gaussianBlur(bump, W, H, sigma);
    const N = bump.length;
    const out = new Float32Array(N);
    for (let i = 0; i < N; i++) {
        out[i] = bump[i] + alpha * (bump[i] - blurred[i]);
    }
    return out;
}

/**
 * 把居中 float bump → uint16 / uint8 整数数组（用于 PNG 编码）。
 * 输入先 clip 到 [-0.5, 0.5]，平移到 [0, 1]，再缩放到目标位深。
 *
 * @param {Float32Array} bump
 * @param {8 | 16} bitDepth
 * @returns {Uint8Array | Uint16Array}
 */
export function quantizeToArray(bump, bitDepth = 16) {
    if (bitDepth !== 8 && bitDepth !== 16) {
        throw new Error(`bitDepth must be 8 or 16, got ${bitDepth}`);
    }
    const N = bump.length;
    const max = bitDepth === 16 ? 65535 : 255;
    const out = bitDepth === 16 ? new Uint16Array(N) : new Uint8Array(N);
    for (let i = 0; i < N; i++) {
        let v = bump[i];
        if (v < -0.5) v = -0.5;
        else if (v > 0.5) v = 0.5;
        const shifted = v + 0.5; // [0, 1]
        let q = (shifted * max + 0.5) | 0;
        if (q < 0) q = 0;
        else if (q > max) q = max;
        out[i] = q;
    }
    return out;
}
